hazelcast/src/main/java/com/hazelcast/cache/impl/CacheEventHandler.java

```diff
@@ -16,93 +16,58 @@
-
-    private boolean invalidationMessageBatchEnabled;
-    private int invalidationMessageBatchSize;
-    private ScheduledFuture cacheBatchInvalidationMessageSenderScheduler;
+    private final Invalidator invalidator;
 
     CacheEventHandler(NodeEngine nodeEngine) {
-        this.logger = nodeEngine.getLogger(getClass());
         this.nodeEngine = nodeEngine;
-        this.partitionService = nodeEngine.getPartitionService();
-        this.metaDataGenerator = new MetaDataGenerator(partitionService.getPartitionCount());
-        HazelcastProperties properties = nodeEngine.getProperties();
-        invalidationMessageBatchEnabled = properties.getBoolean(CACHE_INVALIDATION_MESSAGE_BATCH_ENABLED);
-        if (invalidationMessageBatchEnabled) {
-            invalidationMessageBatchSize = properties.getInteger(CACHE_INVALIDATION_MESSAGE_BATCH_SIZE);
-            int invalidationMessageBatchFreq = properties.getInteger(CACHE_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS);
-            ExecutionService executionService = nodeEngine.getExecutionService();
-            CacheBatchInvalidationMessageSender batchInvalidationMessageSender = new CacheBatchInvalidationMessageSender();
-            cacheBatchInvalidationMessageSenderScheduler = executionService
-                    .scheduleWithRepetition(ICacheService.SERVICE_NAME + ":cacheBatchInvalidationMessageSender",
-                            batchInvalidationMessageSender,
-                            invalidationMessageBatchFreq,
-                            invalidationMessageBatchFreq,
-                            TimeUnit.SECONDS);
+        this.invalidator = createInvalidator();
+    }
+
+    private Invalidator createInvalidator() {
+        HazelcastProperties hazelcastProperties = nodeEngine.getProperties();
+        int batchSize = hazelcastProperties.getInteger(CACHE_INVALIDATION_MESSAGE_BATCH_SIZE);
+        int batchFrequencySeconds = hazelcastProperties.getInteger(CACHE_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS);
+        boolean batchingEnabled = hazelcastProperties.getBoolean(CACHE_INVALIDATION_MESSAGE_BATCH_ENABLED) && batchSize > 1;
+
+        if (batchingEnabled) {
+            return new BatchInvalidator(SERVICE_NAME, batchSize, batchFrequencySeconds, TRUE_FILTER, nodeEngine);
+        } else {
+            return new NonStopInvalidator(SERVICE_NAME, TRUE_FILTER, nodeEngine);
         }
-        LifecycleService lifecycleService = nodeEngine.getHazelcastInstance().getLifecycleService();
-        lifecycleService.addLifecycleListener(new LifecycleListener() {
-            @Override
-            public void stateChanged(LifecycleEvent event) {
-                if (event.getState() == SHUTTING_DOWN) {
-                    invalidateAllCaches();
-                }
-            }
-        });
     }
 
-    public MetaDataGenerator getMetaDataGenerator() {
-        return metaDataGenerator;
+    public Invalidator getInvalidator() {
```

Comment:
All usages of this method seem to be to retrieve the `MetaDataGenerator`. Can't we keep the old method to reduce the number of needed changes? Then it would be an implementation detail how to retrieve the `MetaDataGenerator`.